### ISSUE TITLE
[BUG] Storage SDK hitting OOM with stageBlockWithResponse 12.0.0.preview-3

### ISSUE THREAD

**Issue by vmaheshw** [Azure/azure-sdk-for-java#5607]
**Describe the bug**
@rickle-msft 
Uploading Blocks of size 1-2MB with ~2500 blocks per minute.
I'm hitting OOM in 7 minutes of run.

I'm using async mechanism to upload multiple blocks in parallel. My code snippet is working properly with V10.5.0. For V12, all I did is change from
Flowable<ByteBuffer> outputStream = Flowable.just(ByteBuffer.wrap(localByte, 0, blockSize));
to 
outputStream = Flux.just(ByteBuffer.wrap(localByte, 0, blockSize));

and then calling:
 blobAsyncClient.stageBlockWithResponse(blockIdEncoded, Flux.just(outputStream), blockSize, null).block();

The main classes contributing in HeapDump are:
HeapByteBuffer
ConcurrentHaspMap which contains
FluxJust
FluxMap
FluxDefer

I was able to run for 10+ hours with preview-2 which was using ByteBuf instead of ButeBuffer before hitting another issue.
***Exception or Stack Trace***
Add the exception log and stack trace if available

**To Reproduce**
Just upload lot of blocks using BlockBlobAsyncClient using api StageBlockWithResponse

***Code Snippet***
outputStream = Flux.just(ByteBuffer.wrap(localByte, 0, blockSize));

**Expected behavior**
No OOM and proper cleanup.

**Screenshots**

**Setup (please complete the following information):**
 - IDE : [e.g. IntelliJ]
 - Version of the Library used: V12-preview3

**Additional context**

**Information Checklist**
Kindly make sure that you have added all the following information above and checkoff the required fields otherwise we will treat the issuer as an incomplete report
- [x] Bug Description Added
- [x] Repro Steps Added
- [x] Setup information Added

**Comment by rickle-msft** [comment]
@vmaheshw Thank you for reporting this issue. If possible, could you please offer a more complete code snippet that we can run? Or if not, it would be helpful to know, what is the typical size of `localByte`? Are blocks being staged in a tight loop? Is the app allocating data before each ByteBuffer wrapping or is it pooling? 

@jianghaolu @anuchandy @alzimmermsft can you guys comment on whether was behavior in core that changed between 2 and 3 that could be causing memory leaks or something similar?

**Comment by vmaheshw** [comment]
localByte size is ~1-2MB. See the below snippet on how the ByteBuffer is getting wrapped.


Void uploadBlock(blockId, byte[] localByte) {

CompletableFuture<Void> future = CompletableFuture.runAsync(new Runnable() {
   @Override
  public void run() {
    int blockSize = localByte.length;
    ByteBuffer outputStream = ByteBuffer.wrap(localByte, 0, blockSize);
     blobAsyncClient.stageBlockWithResponse(blockIdEncoded, Flux.just(outputStream), blockSize, null).block();
  }
}, blobThreadPool);

pendingUpload.add(future);
future.handle((aVoid, throwable) -> {
  pendingUpload.remove(future);
});
}

**Comment by alzimmermsft** [comment]
Hi @vmaheshw,

Thanks for submitting this issue, we'll begin attempting to replicate it and find a root cause. Once the root cause is found we'll update the issue with the cause and the solution to the issue.

**Comment by anuchandy** [comment]
@vmaheshw some more data about your scenario will help us debug this. 

so you are using following method to upload one block:

```java
void uploadBlock(blockId, byte[] localByte) {
    CompletableFuture<Void> future = CompletableFuture.runAsync(new Runnable() {
       @Override
       public void run() {
           int blockSize = localByte.length;
           ByteBuffer outputStream = ByteBuffer.wrap(localByte, 0, blockSize);
           blobAsyncClient.stageBlockWithResponse(blockIdEncoded, 
               Flux.just(outputStream), 
               blockSize, null).block();
       }
    }, threadPool);

    completableFutureList.add(future);
    future.handle((aVoid, throwable) -> {
        completableFutureList.remove(future);
    });
}
```

It would be very helpful to see how this method is being used, basically we're looking answers for following questions:

1. What is the size of `threadPool`? is it unbounded? 
2. Does multiple `uploadBlock` method call happens serially or concurrently? if concurrently then is `completableFutureList` a concurrent list?
3. How are you using `uploadBlock` method to upload N blocks? (you mentioned ~2500 blocks/min)
       a]. is it a simple loop with upper bound N where each iteration uploads 1 block ?
       b]. or do you split N  `uploadBlock` calls into M groups where each group uploads N/M blocks?
       c]. or some other logic
       In any case, between single upload or group of uploads, is there a wait or it's a tight loop?

Is it possible to share the usage code which help answering above questions? if code cannot be extracted out from your app then could you answer above questions?

**Comment by vmaheshw** [comment]
@anuchandy:
1. For this exercise, you can assume thread pool size to be 30. I have 4 tasks running each with thread pool size 30.
2. Concurrently. We are not waiting for the upload to finish. As soon as we have data ready to upload, we stageBlock asynchronously.
3. Each iteration is uploading a different block. There are 12 blobs in which these blocks are distributed. But, This issue is possible even if there is 2500 blocks for 1 blob uploaded in 1 minute.
Main thread is processing data and preparing the localByte. Once it is ready to be uploaded, it is passed to threadpool to upload async.

**Comment by vmaheshw** [comment]
@anuchandy @alzimmermsft Any update on this?

**Comment by srnagar** [comment]
I am working with @vmaheshw on Teams to get a repro for this issue. I will update this ticket with my findings.

**Comment by srnagar** [comment]
@vmaheshw I was able to reproduce this issue locally and preview 4 will have the fix for this issue.

TL;DR - The issue occurs only when using a **proxy** with reactor-netty 0.8.3.RELEASE version. 

Preview 3 uses 0.8.3.RELEASE version of reactor-netty. 
Preview 4 uses reactor-netty 0.9.0.RELEASE which has the fix.

**Details**

This issue occurs when HttpClient is configured to use a proxy. In this case, the issue was not reproducible when I was using a default HttpClient but after configuring a proxy for my HttpClient, I saw OOM error. 

The root cause for this issue is that the connection pool was adding a new entry to the internal hashmap for every upload operation and holding on to all the HttpResources. These resources were not garbage collected and eventually resulted in OOM. When using proxy, the hashcode computed for the key was using a `Supplier` instead of the host address and each invocation resulted in a new hashcode being generated. So, the size of `channelPools` map in `PooledConnectionProvider` kept increasing with every `stageBlockWithResponse()` operation.

This issue was reported to the Reactor team - https://github.com/reactor/reactor-netty/issues/566
The fix is available from version 0.8.5.RELEASE. 

Preview 4 release of Azure track 2 SDKs is using 0.9.0.RELEASE version of reactor-netty which will have the fix for the above issue.

**More details for anyone looking to reproduce this.**

**Code snippet**
```java
      HttpClient httpClient = new NettyAsyncHttpClientBuilder()
          .proxy(new ProxyOptions(ProxyOptions.Type.HTTP,
              new InetSocketAddress(hostName, 8080))).build();

      BlobServiceAsyncClient blobServiceAsyncClient = new BlobServiceClientBuilder()
          .endpoint(endpoint)
          .credential(new SharedKeyCredential(accountName, accountKey))
          .httpClient(httpClient)
          .buildAsyncClient();

      BlobAsyncClient blobAsyncClient =
          blobServiceAsyncClient.getBlobContainerAsyncClient(CONTAINER_NAME).getBlobAsyncClient(blobName);
      BlockBlobAsyncClient blockBlobAsyncClient = blobAsyncClient.getBlockBlobAsyncClient();

      byte[] localByte = new byte[2 * 1024 * 1024];
      Arrays.fill(localByte, (byte) 1);
      AtomicInteger count = new AtomicInteger(0);
      List<CompletableFuture<Void>> completableFutureList = new ArrayList<>();

      for (int i = 0; i < 500; i++) {
        ThreadPoolExecutor threadPool = (ThreadPoolExecutor) Executors.newFixedThreadPool(10);
        CompletableFuture<Void> future = CompletableFuture.runAsync(new Runnable() {
          @Override
          public void run() {
            IntStream.generate(() -> ThreadLocalRandom.current().nextInt(0, Integer.MAX_VALUE)).map(val -> {
              int blockSize = localByte.length;
              ByteBuffer outputStream = ByteBuffer.wrap(localByte, 0, blockSize);
              blockBlobAsyncClient
                  .stageBlockWithResponse(Base64.encodeBase64String(String.valueOf(0).getBytes()),
                      Flux.just(outputStream),
                      blockSize, null).block();
              System.out.println("Uploaded a block running on thread " + Thread.currentThread().getName() + " "
                  + "count = " + count.get() + " at " + OffsetDateTime.now());
              return val;
            }).count();
          }
        }, threadPool);
        completableFutureList.add(future);
        future.handle((aVoid, throwable) ->
            {
              System.out.println("Exception " + throwable.getMessage());
              return completableFutureList.remove(future);
            }
        );
      }
      CompletableFuture.allOf(completableFutureList.toArray(new CompletableFuture[completableFutureList.size()]));
```

To create `BlobServiceAsyncClient` without proxy, use this code snippet instead:
```java
BlobServiceAsyncClient blobServiceAsyncClient = new BlobServiceClientBuilder()
          .endpoint(endpoint)
          .credential(new SharedKeyCredential(accountName, accountKey))
          .httpClient(HttpClient.createDefault())
          .buildAsyncClient();
```

State of `channelPools` in `PooledConnectionProvider` class with 0.8.3.RELEASE version
![image](https://user-images.githubusercontent.com/51379715/66361822-27b19a00-e935-11e9-874f-5e29d97c629d.png)

State of `channelPools` in `PooledConnectionProvider` class with 0.9.0.RELEASE version
![image](https://user-images.githubusercontent.com/51379715/66361835-3435f280-e935-11e9-860c-b451c1998165.png)

**Comment by vmaheshw** [comment]
Verified this fix. I'm closing this issue. Thank you @srnagar
